File: src/signal.rs
//! Best-effort POSIX signal identification.
//!
//! We deliberately avoid a libc dependency for this: the set of signals we
//! actually need to name for classification purposes is small and stable.

use serde::{Deserialize, Serialize};

/// Information about a POSIX signal that terminated a process.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SignalInfo {
    /// Raw signal number (e.g. `11` for `SIGSEGV`).
    pub number: i32,
    /// Symbolic name, if recognized (e.g. `"SIGSEGV"`).
    pub name: Option<String>,
    /// Whether the OS reported that a core dump was produced.
    pub core_dumped: bool,
}

impl SignalInfo {
    pub fn new(number: i32, core_dumped: bool) -> Self {
        Self {
            number,
            name: signal_name(number).map(str::to_string),
            core_dumped,
        }
    }
}

/// Maps common POSIX signal numbers to their symbolic names.
///
/// Only the signals relevant to diagnosing native crashes are covered; an
/// unrecognized number yields `None` rather than a guess.
///
/// Numbering here follows macOS/BSD (`sys/signal.h`), which is the
/// development and test platform for this component. The signals that
/// drive classification (`SIGSEGV`=11, `SIGABRT`=6, `SIGFPE`=8, `SIGILL`=4)
/// share the same numbers on Linux; `SIGBUS`, `SIGSYS`, `SIGUSR1`, and
/// `SIGUSR2` do not, so their names may be misreported when TraceCLI is
/// eventually run on Linux. That is a known platform limitation, not a
/// correctness target for this MVP.
pub fn signal_name(number: i32) -> Option<&'static str> {
    match number {
        1 => Some("SIGHUP"),
        2 => Some("SIGINT"),
        3 => Some("SIGQUIT"),
        4 => Some("SIGILL"),
        5 => Some("SIGTRAP"),
        6 => Some("SIGABRT"),
        7 => Some("SIGEMT"),
        8 => Some("SIGFPE"),
        9 => Some("SIGKILL"),
        10 => Some("SIGBUS"),
        11 => Some("SIGSEGV"),
        12 => Some("SIGSYS"),
        13 => Some("SIGPIPE"),
        14 => Some("SIGALRM"),
        15 => Some("SIGTERM"),
        30 => Some("SIGUSR1"),
        31 => Some("SIGUSR2"),
        _ => None,
    }
}

File: src/signal.js
'use strict';

const os = require('node:os');

// Best-effort POSIX signal identification. Node's own `os.constants.signals`
// gives us the *current platform's* correct signal numbers (unlike the
// Rust/C++ ports, which had to hardcode a macOS/BSD table and documented a
// known Linux mismatch for SIGBUS/SIGUSR1/2) -- so this port doesn't
// inherit that limitation.

/** Returns the platform-correct signal number for a name like 'SIGSEGV',
 * or undefined if unknown. */
function signalNumber(name) {
  const num = os.constants.signals[name];
  return typeof num === 'number' ? num : undefined;
}

/** Builds a SignalInfo from a signal *name* as reported by
 * `child_process` (Node reports the name directly, not a raw number).
 * `coreDumped` is best-effort: Node's public API does not expose the
 * WCOREDUMP bit, so this is always `false` -- a documented limitation,
 * not a correctness target. */
function signalInfoFromName(name, coreDumped = false) {
  return {
    number: signalNumber(name) ?? 0,
    name: name ?? null,
    core_dumped: coreDumped,
  };
}

module.exports = { signalNumber, signalInfoFromName };
